Update heading block on ENTER key to create empty paragraph instead of heading

andrewserong opened the pull request:
Resolves #10326

## Description

This PR tweaks the behaviour for the heading block when a user presses ENTER at the beginning of a heading (as mentioned in #10326).

This allows the ENTER key pressed at the beginning of a heading to create an empty paragraph above it. Also, when the ENTER key is pressed at the end of a heading, an empty paragraph block is created below, instead of another heading block.

I realise this could be a matter of personal preference, but to me this feels like a more natural behaviour when writing long-form prose, where the default block would seem to be the paragraph, and I assume a user creating headings would want to switch back to a paragraph block instead of creating multiple heading blocks.

## How has this been tested?
Unit tests and linting is passing locally. Tested in Chrome 69 and Safari 11.1.1 on macOS 10.13.5.

## Screenshots <!-- if applicable -->

![gutenberg-heading-enter](https://user-images.githubusercontent.com/14988353/46581096-82dc5280-ca7d-11e8-88dd-30a300712e0e.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
* Extending the logic of the `onSplit` method in the heading component.

<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

andrewserong committed:
Update heading block behaviour on ENTER key to create empty paragraph instead of heading (#10326)

The onSplit method for the heading block will now replace an empty heading block with an empty paragraph block for the 'before' content.

This allows the ENTER key pressed at the beginning of a heading to create an empty paragraph above it. Also, when the ENTER key is pressed at the end of a heading, an empty paragraph block is created below, instead of another heading block. The RichText component now also calls event.stopImmediatePropagation() before splitContent() to prevent TinyMCE from throwing an error. This avoids a similar error as in #4580.

gziolo commented on packages/editor/src/components/rich-text/index.js:
```diff
@@ -639,6 +639,11 @@ export class RichText extends Component {
+				// calling onReplace() will destroy the editor, so it's
+				// important that we stop other handlers (e.g. ones
+				// registered by TinyMCE) from also handling this event.
+				event.stopImmediatePropagation();
```
Looking at failing e2e tests it seems like this isn't a change that would be expected.

andrewserong replied:
Removing this line has fixed the failing e2e tests.

iseulde commented:
Worth noting that this bug is also addressed in #11005, but it's cool if I can be done without it.

gziolo commented:
@andrewserong do you plan to continue working on this PR? It looks like the changes which @iseulde referenced are much broader but you would have to validate yourself how much overlap there is.

andrewserong commented:
@gziolo apologies for the late reply — I'll take a look over this PR again later on today. I've just rebased master into the branch (since the PR is pretty old now), and will compare against the PR @iseulde mentioned.

andrewserong committed:
Re-order onReplace and insertBlocksAfter for heading unstableOnSplit

Since my last commit, the behaviour of onReplace seems to have changed a little bit. This commit ensures that an empty paragraph is added before a heading if a heading is 'split' with no before text, but with some after text (e.g. if the carat is at the beginning of a heading and the user presses ENTER)

Also removed unnecessary `event.stopImmediatePropagation()`.

andrewserong commented:
@gziolo and @iseulde — I've fixed up this PR, updating the order of `onReplace` and `insertBlocksAfter` (looks like the behaviour of onReplace might have changed slightly since I initially submitted the PR).

Looks like the tests are passing now, but I'm also aware that I haven't added e2e tests for this specific behaviour. It looks like particular blocks aren't targeted in the existing e2e tests, so I wasn't sure if behaviour specific to the heading block should be covered or not?

The other consideration is that this PR works with the current `unstableOnSplit` method, but will be superseded by the changes @iseulde has made to `onSplit`. If you'd like a quick fix for this bug, I'm happy to do any further changes you might like on this PR, otherwise, also happy for you to close it and wait for the revised onSplit prop in #11005.

iseulde commented:
I think it's ok to add the fix, but we should definitely add e2e tests. :)

andrewserong committed:
Add e2e tests to cover onSplit behaviour for headings (#10326)

This ensures that if a user types a heading, then moves the caret to the beginning of the line and presses ENTER, that a paragraph block is created above. Also ensures that when a user presses ENTER at the end of a heading, a paragraph block is created below.

andrewserong commented:
Thanks @iseulde! — I've added e2e tests now to cover the initial bug report in #10326, so the issue should be adequately covered when this fix is superseded by your revised onSplit prop. cc: @gziolo

gziolo commented:
Nice addition with tests. Let’s merge it as is after some testing in that case 👍

iseulde commented on packages/e2e-tests/specs/splitting-merging.test.js:
```diff
@@ -210,4 +210,24 @@ describe( 'splitting and merging blocks', () => {
 		expect( isInDefaultBlock ).toBe( true );
 	} );
+
+	it( 'should create a paragraph block above when pressing enter at start of heading', async () => {
```
Could we move these files to `blocks/heading.test.js`, as this is specific behaviour for headings?

andrewserong replied:
That makes sense, I'll move it over.

iseulde commented on packages/e2e-tests/specs/splitting-merging.test.js:
```diff
@@ -210,4 +210,24 @@ describe( 'splitting and merging blocks', () => {
+		await page.keyboard.press( 'ArrowUp' );
+
+		const activeElementClassList = await page.evaluate( () => document.activeElement.classList );
+		expect( Object.values( activeElementClassList ) ).toContain( 'wp-block-paragraph' );
```
What is the arrow up and active class check testing for? Is checking the content not enough?

andrewserong replied:
I was initially concerned with checking the active element of the caret position, but I think you're right, if we're checking the content, the active class check is overkill, I'll remove it.

andrewserong committed:
Move heading e2e tests to the heading block, remove extra active class check (#10326)

andrewserong commented:
Thanks for the feedback @iseulde! I've moved the tests up to the heading block tests, and removed the extra active class check.

iseulde commented:
Something strange happens when merging back:

* Put the caret in front of the heading text.
* Press enter. An empty paragraph is inserted before the heading block. The caret stays in place. All good.
* Now press backspace to merge the blocks.
  * Expected: remove the empty paragraph block and the caret should stay in place.
  * Actual: removes the paragraph block, but also transforms the heading block into a paragraph block and the caret is put at the end...

iseulde commented:
I could live with it being transformed to a paragraph block, but the caret moving to the end is very unexpected. This doesn't happen in `master`.

iseulde commented on packages/e2e-tests/specs/blocks/heading.test.js:
```diff
@@ -5,6 +5,8 @@ import {
+	pressKeyTimes,
 } from '@wordpress/e2e-test-utils';
 
 describe( 'Separator', () => {
```
Not introduced in this PR, but this title is not correct. 😅

andrewserong commented:
> I could live with it being transformed to a paragraph block, but the caret moving to the end is very unexpected. This doesn't happen in `master`.

Oh, thanks for spotting that! Yeah, that's not meant to happen. I'll take another look and see if I can work out what's going on...

andrewserong commented:
I'm a little bit stumped by this issue — it feels to me like it's got to do with how `onReplace` behaves. I've run out of time for working on it tonight, but I'll take another look over the weekend.

andrewserong commented:
I've done a bit more investigating, and I think there might be an issue that exists on Master that affects the behaviour here. If we have an empty paragraph block that already exists, and press backspace at the beginning of a heading block, the paragraph block above is removed, and the caret sits at the beginning of the heading as we would expect. It looks like this issue only occurs on paragraph blocks that have been created above _by_ the heading block when it's split. So it looks like there's something different about the block that was 'created' by the heading block onSplit than a standard paragraph block — even though the code editor view of the block remains the same.

The issue that I noticed on Master occurs if you create a heading, then move the caret to the beginning of the line, and then press ENTER. The heading moves down to a paragraph block as we would expect, but the empty heading block now has two `<h2>` tags in it, which isn't usually noticeable, but if you're using the 2019 theme, you can see two horizontal lines in the Heading block where there should be just one:

![image](https://user-images.githubusercontent.com/14988353/52909286-83fa7780-32da-11e9-8d73-38d07525eca9.png)

My hunch at the moment is that whatever it is about this block that gets created from the Heading block when a user hits ENTER is causing it to not do what we expect it to when we want to merge those blocks again. If that makes sense!

Since this issue is affecting this PR, I'm happy to keep investigating when I have time, but I don't want to hold anything up if there's a desire to get this one merged in. Let me know, and I'm happy to open a new issue for this.

ellatrix commented:
@andrewserong Thank you for the work you've done here! I've refactored selection state in #14640 and now #14765 will revise the `onSpit` prop entirely and fix this issue as well, so I'm going to close this PR.

andrewserong commented:
@ellatrix That's great to hear, thanks for letting me know!
